diesel/src/pg/backend.rs

```diff
@@ -94,7 +94,7 @@ impl<'a> HasRawValue<'a> for Pg {
 
 impl TypeMetadata for Pg {
     type TypeMetadata = PgTypeMetadata;
-    type MetadataLookup = PgMetadataLookup;
+    type MetadataLookup = dyn PgMetadataLookup;
```

Comment:
I agree we shouldn't need dynamic dispatch here.
If it turns out we need it here, this would mean that the `MetadataLookup` is actually not a property of the `Backend`, but instead a property of the connection.

However, I beleive that the way we lookup metadata is actually a property of the backend, but that we need a way to represent how we would lookup stuff with a given connection.
It looks to me like this could be the model:
```rust
pub struct PgMetadataLookup<C: Connection<Backend=Pg>> {
    conn: C,
}
// and basically same `impl`s as now

pub struct PgMetadataLookupKind;
// "Kind" because https://en.wikipedia.org/wiki/Kind_(type_theory)

pub trait MetadataLookupKind<C> {
    type Out;
}
impl<C: Connection<Backend=Pg>> MetadataLookupKind<C> for PgMetadataLookupKind {
    type MetadataLookup = PgMetadataLookup<C>;
}
impl TypeMetadata for Pg {
    type TypeMetadata = PgTypeMetadata;
    type MetadataLookup = PgMetadataLookupKind;
}

pub trait HasSqlType<ST>: TypeMetadata {
    fn metadata<C>(
        lookup: &<Self::MetadataLookup as MetadataLookupKind<C>>::MetadataLookup,
    ) -> Self::TypeMetadata
    where Self::MetadataLookup: MetadataLookupKind<C>;
}

impl HasSqlType<Int4range> for Pg {
    fn metadata<C>(_: &PgMetadataLookup<C>) -> PgTypeMetadata
    where C: Connection<Backend=Self>,
    {
        PgTypeMetadata::new(3904, 3905)
    }
}
```

Reply:
Is it ok to force this even more complex interface on all `Backend` and `HasSqlType` implementations just so one `Backend` can have dynamic types?

Reply:
Also I just realized that this interface would tie the `Pg` backend to the sync `Connection` api. This would make it even harder to make an async api. This is kind of alright because at least on tokio you could use `block_in_place` to create a sync api and it would only be called once per user defined type.

Reply:
What is the performance impact of dynamic dispatch here? I couldn't actually find and really general ballpark except here:  https://github.com/rust-lang/rfcs/pull/1603#issuecomment-216861906. Also I kind of doubt the that the look up would be inlined or in the hot path of ast. It is only ever being constructed once per query and it would only be called a couple of times max in a query as most types don't actually call any of the trait's methods.

Reply:
It's always hard to guess what impact a certain change has, therefore it's preferred to just measure that. Therefore I've put together some small piece of code demonstrating the worst case scenario:

```rust
    #[derive(crate::sql_types::SqlType, QueryId, Clone, Copy, Debug)]
    #[postgres(type_name = "my_enum")]
    pub struct MyEnumSqlType;

    #[allow(warnings)]
    #[test]
    fn bench_oid_lookup() {
        use crate::expression::AsExpression;
        use crate::prelude::*;
        use crate::query_builder::QueryFragment;
        use crate::serialize::ToSql;
        use crate::sql_types::SqlType;

        table! {
            use super::MyEnumSqlType;
            use crate::sql_types::*;
            test_table(id) {
                id -> Integer,
                my_enum -> MyEnumSqlType,
            }
        }

        #[derive(AsExpression, FromSqlRow, Copy, Clone, Debug)]
        #[sql_type = "MyEnumSqlType"]
        enum MyEnum {
            Value,
        }

        impl ToSql<MyEnumSqlType, Pg> for MyEnum {
            fn to_sql<W: std::io::Write>(
                &self,
                out: &mut crate::serialize::Output<W, Pg>,
            ) -> crate::serialize::Result {
                unimplemented!()
            }
        }

        let conn = connection();

        conn.begin_test_transaction().unwrap();

        conn.execute("CREATE TYPE my_enum AS ENUM ('value')")
            .unwrap();

        // setup till here

        let query =
            crate::insert_into(test_table::table)
                .values(vec![test_table::my_enum.eq(MyEnum::Value); 65000]);

        let mut bind_collector = RawBytesBindCollector::<Pg>::new();
        query
            .as_query()
            .collect_binds(&mut bind_collector, PgMetadataLookup::new(&conn))
            .unwrap();
    }
```

The underlying point here is that while constructing your prepared statement we will call `HasSqlType::metadata` for each bind value. That query has 65k bind values for a custom enum type, which require a dynamic lookup every time. Therefore if we go the full trait object way, this should demonstrate the worst case impact of such a change. Would it possible for you to run that for your version and the existing implementation to gather some times? If we have such timings we can really discuss which way should be chosen and which is overly complicated (or introduces to high performance costs for some use cases.)

Reply:
It's probably true this has a rather low perf impact, especially if done after the caching (as we would make a query there anyway).
If we want to allow different connections to be used without introducing genericity, I beleive it's better to introduce the dyn dispatch after the caching, as suggested at https://github.com/diesel-rs/diesel/pull/2639#discussion_r570364515.
In which case, it looks like an interface doing that could be:
```rust
#[repr(transparent)]
pub struct PgMetadataLookup {
	conn: dyn PgConnectionForMetadataLookup,
}
trait PgConnectionForMetadataQuery {
	fn lookup_type(&self, cache_key: &PgMetadataCacheKey<'_>) -> QueryResult<InnerPgTypeMetadata>;
}
impl<C: Connection<Backend = Pg>> PgConnectionForMetadataQuery for C {
	fn lookup_type(&self, cache_key: &PgMetadataCacheKey<'_>) -> QueryResult<InnerPgTypeMetadata> {
		// same impl as now in the private function of the module
		todo!()
	}
}
pub(crate) trait PgConnectionWithMetadataCache {
	fn metadata_cache(&self) -> &PgMetadataCache;
}
trait PgConnectionForMetadataLookup:
	PgConnectionWithMetadataCache + PgConnectionForMetadataQuery
{
}
impl<T: PgConnectionWithMetadataCache + PgConnectionForMetadataQuery> PgConnectionForMetadataLookup
	for T
{
}

impl PgMetadataLookup {
	#[allow(clippy::new_ret_no_self)]
	pub(crate) fn new<C>(conn: &dyn PgConnectionForMetadataLookup) -> &Self {
		unsafe { &*(conn as *const dyn PgConnectionForMetadataLookup as *const PgMetadataLookup) }
	}

	pub fn lookup_type(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata {
		// handle caching and fallback to self.conn_for_lookup.lookup_type()
		todo!()
	}
}
```
This would also solve the issue underlined at https://github.com/diesel-rs/diesel/pull/2639#discussion_r570400958.

Reply:
I would consider `#[repr(transparent)]` with a constructor that just converts one pointer to the other to be an anti-pattern. Just use the original type, because that is what you are doing. By the way, `dyn Trait` is a type.

Reply:
@Ten0 That implementation will also require a dynamic dispatch per each call to `lookup`.

Otherwise I really want to see some numbers before deciding whats the best solution here.

Reply:
> I would consider `#[repr(transparent)]` with a constructor that just converts one pointer to the other to be an anti-pattern. Just use the original type, because that is what you are doing. By the way, `dyn Trait` is a type.

It looks like this is the newtype pattern (https://doc.rust-lang.org/book/ch19-04-advanced-types.html#using-the-newtype-pattern-for-type-safety-and-abstraction). It allows to not expose the inner type, which we're happy of in our case, as this avoids making the mistake to query a type through the non-cached interface (https://github.com/diesel-rs/diesel/pull/2639#discussion_r570400958).

Now this pattern is annoying to use in our case, as we'd like to build it safely, but to build it normally it would have to hold the reference, which would make it relative to a lifetime, and then it couldn't be used as associated type for the `TypeMetadata` trait, because we don't have GATs.

Now this use-case is not so uncommon, and there's even a library with > 1M downloads that provides an interface for doing ref-to-ref newtype pattern safely.
However I feel like for now, https://github.com/dtolnay/ref-cast/issues/3 not being solved makes it non satisfactory for use on public types within diesel (such as `PgMetadataLookup`).

Reply:
> @Ten0 That implementation will also require a dynamic dispatch per each call to `lookup`.

Oh, because it needs to get its ref to the cache, right :(

Reply:
After thinking about this, I thought maybe the reason for this weird type conversion is to limit the api. Is that why?

Reply:
Yes restricting the api to a single method is one reason for the current design.

Reply:
Is it possible to make the type a function? 

Also with some experimentation, I found out you can impl methods and traits for `dyn Trait`. I thought this would solve our problem but you still need access to the cache somehow.

Reply:
Ok, we could do a function pointer. But this would be hard for `PgConnection` to implement and might be another form of dynamic dispatch.

Reply:
My thoughts right now is that the constructor pattern is probably the only way of avoiding dynamic dispatch. I do think that we could design it without requiring implementing `Connection<Backend=Pg>`. It could be a trait with just the required methods like `get_cache` and `lookup`.

Reply:
Just as a sanity check, I benched the worst case that you provided. The difference was about 4% on collecting the bindings. There was no measurable difference in actually executing the query. Collecting the bindings took about 4ms and executing the query took about 52ms.

This isn't the best benchmark because I cut some corners to save time.
- `ToSql` implmentation for `MyEnum` is just `Ok(diesel::serialize::IsNull::Yes)`
- I didn't create the test table so the query just fails

I just wanted a ball park estimate.

Reply:
Here is my proposed solution.

```rust
impl TypeMetadata for Pg {
    type TypeMetadata = PgTypeMetadata;
    type MetadataLookup = dyn PgTypeMetadataLookup;
}

trait PgTypeMetadataLookup {
    fn lookup(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata;
}

impl<T> PgTypeMetadataLookup for T where T: PgCachedTypeMetadataLookup {
    fn lookup(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata {
        todo!();
    }
}

trait PgCachedTypeMetadataLookup {
    fn get_cache(&self) -> PgTypeMetadataCache;
    fn uncached_lookup(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata;
}


impl PgCachedTypeMetadataLookup for PgConnection {
    fn get_cache(&self) -> PgTypeMetadataCache {
        todo!();
    }
    
    fn uncached_lookup(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata {
        todo!();
    }
}
```

Here are the reasons I think this solution is good:
- This only has one dynamic dispatch per call to look up a user defined type. I believe this will not negatively impact the performance on actual query execution. 
- It leaves the `HasSqlType` trait unchanged. This is good because most implementations never use `MetadataLookup` and it feels odd to make the trait some complicated for the benefit of a few implementations.
- It implements caching once and other connections can just implement `PgCachedTypeMetadataLookup` to get caching for free. 
- It limits the API of `PgTypeMetadataLookup` to one method. This means implementations of `HasSqlType` can't bypass caching.
- It allows connections that want to be `Send + Sync` to use their own caching mechanism, which I believe future async connections will want to do.

Reply:
This looks really good, with two minor adjustment: 
* I believe its possible to implement the uncached lookup generically for all connections. Likely we need to have a 3 trait setup for that. One to get the cache, which should be implemented manually by each connection, one to perform the uncached lookup, which is generically implemented for all connections which set backend to Pg and one that is a supertrait of both, exporting the public lookup method. That is implemented for all connections implementing the other two trait. I propose to use different traits for all of that to just separate which things need to be public and which do not need to be public.
* We need to return some reference to the cache, not the cache itself.

Reply:
That sounds good! That means connections just have to implement the cache trait and than everything else is implemented for them. Really good idea. I will start working on this.

Reply:
Actually this can be done with only two traits if you want.
```rust
impl TypeMetadata for Pg {
    type TypeMetadata = PgTypeMetadata;
    type MetadataLookup = dyn PgTypeMetadataLookup;
}

trait PgTypeMetadataLookup {
    fn lookup(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata;
}

impl<T> PgTypeMetadataLookup for T where T: GetPgTypeMetadataCache + Connection<Backend=Pg> {
    fn lookup(&self, type_name: &str, schema: Option<&str>) -> PgTypeMetadata {
        todo!();
    }
}

trait GetPgTypeMetadataCache {
    fn get_cache(&self) -> &PgTypeMetadataCache;
}


impl GetPgTypeMetadataCache for PgConnection {
    fn get_cache(&self) -> &PgTypeMetadataCache {
        todo!();
    }
}
```

Reply:
That is also fine :+1:

Reply:
Weird. I benched marked this version and now it is about ~12% slower than master (from 4.3469 ms to 4.8660 ms). To me it should be the exact same, maybe it is just my computer is acting differently today. I could run all three versions back to back if y'all want me to. 

I also found that for query execution is seems to not affect anything. Each time I run the benchmark it is either randomly 11% faster or slower even when running the same benchmark. I think the network is throwing in too much variation for this to be a reliable benchmark. It could also be docker. I am running the database in docker using the docker-compose configuration provided by diesel.

The execution benchmark take about 54 ms and the difference between the binding collection is about 0.5191 ms. So it might cause a 1% slow down in the query execution. But my execution benchmark fails at the point of validating the query on the Postgres server so I would expect the query would take even longer if it actually inserted all the values into the database.

Any thoughts?

Reply:
If you push your last version I will try to do the benchmarks tomorrow.

Reply:
So I've run the benchmark on my machine and to my surprise the dynamic dispatch version is even faster then the old implementation :flushed: 
```
group         dynamic_dispatch                       static_dispatch
-----         ----------------                       ---------------
lookup_oid    1.00      6.0±0.10ms        ? B/sec    1.13      6.8±0.19ms        ? B/sec
```

Criterion seems to be quite sure about that, at least for me. So that's a clear :+1: for me for going down that way.
@Ten0 I've used [this](https://gist.github.com/weiznich/11ac6b7f52d896d8852f65f9fb94efff) benchmark. Feel free to run it on your own, for the case I've screwed something up.